image-registry/automate-pruning.md

```diff
@@ -0,0 +1,266 @@
+   4. Resources for the `CronJob` pod template
+
+These components should be created/reconciled when the image registry operator
+is in the `Managed` state. When the image registry operator is in the `Removed`
+state, the above components should be deleted from the cluster if present.
+
+### Risks and Mitigations
+
+#### Resource consumption
+
+Risk: Image pruning requires siginficant memory to compute the blob layer graph
+needed to identify "prunable" layers.
+
+Mitigation: API will allow cluster admins to adjust the resources allocated to
+the `CronJob`.
+
+Risk: Image pruning requires too much memory to be scheduled to a node
+Risk: Image pruning exhausts resources used for application workloads
+
+Mitigation: API will allow admins to use standard tools to let pruning be run
+on desired machine pools (`affinity`, `tolerations`, `nodeSelector`)
+
+##### Concurrent pruning jobs
+
+Risk: Pruning jobs cannot run concurrently.
+Risk: Pruning job durations can exceed the interval set by the `CronJob`
+schedule.
+
+Mitigation: The `CronJob` for pruning must be configured with
+`concurrencyPolicy: Forbid`
+
+#### Removal of past pruning jobs
+
+Risk: Runs of the pruning job create another set of objects in etcd that need
+to be pruned.
+
+Mitigation: `CronJob` specifies a history limit for successful and failed jobs.
+This will be a tunable parameter via the API.
+
+#### Pruning of in-use images
+
+Risk: Pruner is not aware of a particular image, resulting in it being pruned
+prior to use.
+
+Mitigation: None at present - this is an inherent problem within the current
+pruning process.
+
+## Design Details
+
+### API
+
+Proposed changes to the image registry operator's config spec:
+
+```yaml
+spec:
+  ...
+  pruning:
+    schedule: "*/1 * * * *"
+    keepTagRevisions: 3
+    keepYoungerThan: 60m
+    resources: {}
+    affinity: {}
+    nodeSelector: {}
+    tolerations: {}
+    history:
+      successfulJobsHistoryLimit: 3
+      failedJobsHistoryLimit: 1
+```
+
+1. `schedule:` Cron formatted schedule, defaults to daily at midnight for new
+clusters. If empty, job is suspended.
+2. `keepTagRevisions:` number of revisions per tag to keep. Default 3.
+3. `keepYoungerThan:` retain images younger than this duration. Default 60m.
+4. `resources:` standard Pod resource requests and limits.
+5. `affinity:` standard Pod affinity
+6. `nodeSelector:` standard Pod node selector
+7. `tolerations:` standard Pod tolerations
+8. `history:` sets the job history to be retained on the cluster
+   1. `successfulJobsHistoryLimit:` maximum number of successful jobs to retain
+   2. `failedJobsHistoryLimit:` maximum number of failed jobs to retain
+
+### Fixed Configuration
+
+The following settings for the pruning `CronJob` cannot be configured:
+
+1. `concurrencyPolicy: Forbid` - this prevents concurrent cron jobs from
+running.
+2. `suspend` - based on the provided `schedule`. If empty `suspend = true`.
```

Comment:
We can just delete or skip creation of CronJob if the pruner is disabled.

Reply:
I'm less a fan of that - doing so would remove any retained history from the child `Job`/`Pod` objects.